### PR TITLE
[diffusers] support dynamic batch size

### PR DESCRIPTION
**Pull request by hjchen2**
Depends on this PR https://github.com/siliconflow/oneflow/pull/93, so should update oneflow to the latest version after 2024.1.23.

### CODE REVIEW

File: src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py [hunk 1]
```diff
@@ -684,15 +680,11 @@ def forward(
                     )
                     inner_dim = hidden_states.shape[1]
                     #hidden_states = hidden_states.permute(0, 2, 3, 1).reshape(batch, height * width, inner_dim)
```
Comment (strint) [inline src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py:682]: 可以给改动的地方注释下，方便后面找改动的地方：

# Rewrite for onediff dynamic shape

File: src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py [hunk 2]
```diff
@@ -275,15 +275,11 @@ def forward(
                     hidden_states = self.proj_in(hidden_states)
                     inner_dim = hidden_states.shape[1]
                     # hidden_states = hidden_states.permute(0, 2, 3, 1).reshape(batch, height * width, inner_dim)
-                    hidden_states = hidden_states.permute(0, 2, 3, 1).reshape(
-                        batch, -1, inner_dim
-                    )
+                    hidden_states = hidden_states.permute(0, 2, 3, 1).flatten(1, 2)
```
Comment (strint) [inline src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py:278]: 看起来主要是用  flatten 来解决 reshape 写死 batch 这个 attr 的问题

File: src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+                dtype = torch.float32 if is_mps else torch.float64
+            else:
+                dtype = torch.int32 if is_mps else torch.int64
+            timesteps = torch.tensor([timesteps], dtype=dtype, device=sample.device)
+        elif len(timesteps.shape) == 0:
+            timesteps = timesteps[None].to(sample.device)
+
+        # broadcast to batch dimension in a way that's compatible with ONNX/Core ML
+        # timesteps = timesteps.expand(sample.shape[0])
+        timesteps = torch._C.broadcast_dim_like(timesteps, sample, dim=0)
```
Comment (strint) [inline src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py:155]: 涉及  broadcast_dim_like 的使用，只有这一个地方对吧

File: src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py (hunk 2 above)
Comment (hjchen2) [reply src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py:282]: 是的

File: src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py (hunk 3 above)
Comment (hjchen2) [reply src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py:156]: 是的

File: src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py (hunk 1 above)
Comment (hjchen2) [reply src/infer_compiler_registry/register_diffusers/transformer_2d_oflow.py]: done

**Review by strint — APPROVED** [review]

**Comment by menahem121** [comment]
python text_to_image_sdxl.py 
libibverbs not available, ibv_fork_init skipped
Loading pipeline components...: 100%|████████████████████████████████████████████████████████████████████████████| 7/7 [00:00<00:00,  8.77it/s]
Compiling unet with oneflow.
Compiling vae with oneflow.
Warmup with running graphs...
  0%|                                                                                                                   | 0/30 [00:00<?, ?it/s][ERROR](GRAPH:OneflowGraph_0:OneflowGraph) building graph got error.
ERROR [2024-01-24 17:23:05] - Exception in __call__: e=AttributeError("module 'oneflow._C' has no attribute 'broadcast_dim_like'")
[ERROR](GRAPH:OneflowGraph_1:OneflowGraph) building graph got error.
  0%|                                                                                                                   | 0/30 [00:03<?, ?it/s]
Traceback (most recent call last):
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 172, in wrapper
    return func(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/utils/graph_management_utils.py", line 69, in wrapper
    ret = func(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 257, in __call__
    output = dpl_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 295, in __call__
    self._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 859, in _compile
    return self._dynamic_input_graph_cache._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/cache.py", line 121, in _compile
    return graph._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 863, in _compile
    return self._compile_new(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 882, in _compile_new
    _, eager_outputs = self.build_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 1424, in build_graph
    outputs = self.__build_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 1572, in __build_graph
    outputs = self.build(*lazy_args, **lazy_kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 353, in build
    return self.model(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py", line 156, in forward
    timesteps = torch._C.broadcast_dim_like(timesteps, sample, dim=0)
AttributeError: module 'oneflow._C' has no attribute 'broadcast_dim_like'. Did you mean: 'broadcast_like'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/onediff/examples/text_to_image_sdxl.py", line 75, in <module>
    image = base(
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py", line 814, in __call__
    noise_pred = self.unet(
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/utils/args_tree_util.py", line 34, in wrapper
    output = func(cls, *mapped_args, **mapped_kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 178, in wrapper
    return func(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/utils/graph_management_utils.py", line 69, in wrapper
    ret = func(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 257, in __call__
    output = dpl_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 295, in __call__
    self._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 859, in _compile
    return self._dynamic_input_graph_cache._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/cache.py", line 121, in _compile
    return graph._compile(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 863, in _compile
    return self._compile_new(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 882, in _compile_new
    _, eager_outputs = self.build_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 1424, in build_graph
    outputs = self.__build_graph(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/graph.py", line 1572, in __build_graph
    outputs = self.build(*lazy_args, **lazy_kwargs)
  File "/home/ubuntu/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 353, in build
    return self.model(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/proxy.py", line 188, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/home/ubuntu/onediffenv/lib/python3.10/site-packages/oneflow/nn/graph/proxy.py", line 238, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/home/ubuntu/onediff/src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py", line 156, in forward
    timesteps = torch._C.broadcast_dim_like(timesteps, sample, dim=0)
AttributeError: module 'oneflow._C' has no attribute 'broadcast_dim_like'. Did you mean: 'broadcast_like'?

**Comment by strint** [comment]
@menahem121   cc @hjchen2 

> broadcast_dim_like

This is a new op that was just added into oneflow. So maybe you are using an older oneflow which do not have this op.

What is your oneflow version?  ( You can use this to get: `python -m oneflow --doctor`)

**Pull request merged**